lib.rs

```diff
@@ -605,6 +605,42 @@ impl<A: Array> SmallVec<A> {
             }
         }
     }
+
+    /// Retains only the elements specified by the predicate.
+    ///
+    /// In other words, remove all elements `e` such that `f(&e)` returns `false`.
+    /// This method operates in place and preserves the order of the retained
+    /// elements.
+    pub fn retain<F: FnMut(&A::Item) -> bool>(&mut self, mut f: F) {
+        use std::mem::{forget, swap};
+        match self.data {
+            Inline { ref mut array } => {
+                use std::ptr::read;
+                let mut del = 0;
+                unsafe {
+                    for i in 0..self.len {
+                        if !f(&*array.ptr().offset(i as isize)) {
+                            read(array.ptr().offset(i as isize));
+                            del += 1;
+                        } else if del > 0 {
+                            swap((array.ptr().offset((i - del) as isize) as *mut  A::Item).as_mut().unwrap(), (array.ptr().offset(i as isize) as *mut  A::Item).as_mut().unwrap());
```

Comment:
Indeed if we do that and avoid the `ptr::read`, calling `truncate()` at the end, we can write this just with safe code, right?